### PR TITLE
[Feature] Enable parameter reset in loss module

### PR DESCRIPTION
**Pull request by BY571**
## Description

Allows to reset the parameters in the loss module.

### CODE REVIEW

File: torchrl/objectives/common.py [hunk 1]
```diff
@@ -363,6 +364,35 @@ def reset(self) -> None:
...
+        module_names: Optional[List[Parameter]] = None,
+        init_func: Optional[Callable] = None,
+    ) -> None:
+        """Reset the parameters of the specified modules.
+
+        Args:
+            module_names (Optional[List[Parameter]]): A list of module names to reset the parameters for.
+                If None, all modules with names ending in "_params" will be reset.
+            init_func (Optional[Callable]): A function to initialize the parameters.
+                If None, the parameters will be initialized with uniform random values between -1 and 1.
```
Comment (vmoens) [inline torchrl/objectives/common.py:378]: Seems very unlikely that anyone would want to use that init IMO. Shouldn't we use the init method from the corresponding nn.Module if there is?

File: torchrl/objectives/common.py [hunk 2]
```diff
@@ -363,6 +364,35 @@ def reset(self) -> None:
         # mainly used for PPO with KL target
         pass
 
+    def reset_parameters(
+        self,
+        module_names: Optional[List[Parameter]] = None,
+        init_func: Optional[Callable] = None,
```
Comment (vmoens) [inline torchrl/objectives/common.py:370]:
```suggestion
        init_func: Callable[[torch.Tensor], None] | None = None,
```

File: torchrl/objectives/common.py [hunk 3]
```diff
@@ -363,6 +364,35 @@ def reset(self) -> None:
         # mainly used for PPO with KL target
         pass
 
+    def reset_parameters(
+        self,
+        module_names: Optional[List[Parameter]] = None,
```
Comment (vmoens) [inline torchrl/objectives/common.py:369]:
```suggestion
        module_names: List[Parameter] | None = None,
```

File: torchrl/objectives/common.py [hunk 4]
```diff
@@ -363,6 +364,35 @@ def reset(self) -> None:
...
+        init_func: Optional[Callable] = None,
+    ) -> None:
+        """Reset the parameters of the specified modules.
+
+        Args:
+            module_names (Optional[List[Parameter]]): A list of module names to reset the parameters for.
```
Comment (vmoens) [inline torchrl/objectives/common.py:375]:
```suggestion
            module_names (list of nn.Parameter, optional): A list of module names to reset the parameters for.
```

File: torchrl/objectives/common.py [hunk 5]
```diff
@@ -363,6 +364,35 @@ def reset(self) -> None:
...
+                If None, all modules with names ending in "_params" will be reset.
+            init_func (Optional[Callable]): A function to initialize the parameters.
```
Comment (vmoens) [inline torchrl/objectives/common.py:377]:
```suggestion
            init_func (Callable[[torch.Tensor], None]): A function to initialize the parameters.
```

File: torchrl/objectives/common.py [hunk 6]
```diff
@@ -363,6 +364,35 @@ def reset(self) -> None:
...
+                If None, the parameters will be initialized with uniform random values between -1 and 1.
+        """
+        if module_names is None:
+            params_2_reset = [
+                name for name in self._modules.keys() if name.endswith("_params")
+            ]
+        else:
+            params_2_reset = [name + "_params" for name in module_names]
+
+        def _reset_params(param):
```
Comment (vmoens) [inline torchrl/objectives/common.py:387]: Having one single reset function will be hard to handle, we need a way to tie the reset function and the module.

**Review by vmoens — COMMENTED** [review]
Thanks for this!

We'll need tests for the feature.

How do we handle the target parameters?

Wouldn't something like this be a bit more robust?

```python
from torchrl.objectives import DQNLoss
from torchrl.modules import QValueActor
from torch import nn

module = nn.Sequential(nn.Linear(1, 64), nn.ReLU(), nn.Linear(64, 64))

value_net = QValueActor(module=module, action_space="categorical")
loss = DQNLoss(value_network=value_net, action_space="categorical")

with loss.value_network_params.to_module(loss.value_network):
    loss.apply(lambda module: module.reset_parameters() if hasattr(module, "reset_parameters") else None)
```

**Review by vmoens — COMMENTED** [review]
Thanks for this!

We'll need tests for the feature.

How do we handle the target parameters?

Wouldn't something like this be a bit more robust?

```python
from torchrl.objectives import DQNLoss
from torchrl.modules import QValueActor
from torch import nn

module = nn.Sequential(nn.Linear(1, 64), nn.ReLU(), nn.Linear(64, 64))

value_net = QValueActor(module=module, action_space="categorical")
loss = DQNLoss(value_network=value_net, action_space="categorical")

with loss.value_network_params.to_module(loss.value_network):
    loss.apply(lambda module: module.reset_parameters() if hasattr(module, "reset_parameters") else None)
```

**Comment by BY571** [comment]
> Thanks for this!
> 
> We'll need tests for the feature.
> 
> How do we handle the target parameters?
> 
> Wouldn't something like this be a bit more robust?
> 
> ```python
> from torchrl.objectives import DQNLoss
> from torchrl.modules import QValueActor
> from torch import nn
> 
> module = nn.Sequential(nn.Linear(1, 64), nn.ReLU(), nn.Linear(64, 64))
> 
> value_net = QValueActor(module=module, action_space="categorical")
> loss = DQNLoss(value_network=value_net, action_space="categorical")
> 
> with loss.value_network_params.to_module(loss.value_network):
>     loss.apply(lambda module: module.reset_parameters() if hasattr(module, "reset_parameters") else None)
> ```

I like the solution! But we are accessing the parameters directly in your example so we would need to define a reset function manually, which I think is perfectly fine because then the user has to decide the way how to reset weights and biases:



```python
def reset_parameters(params):
    """ User specified resetting function depending on their needs for initialization """
    if len(params.shape) > 1:
        # weights
        nn.init.xavier_uniform_(params)
    elif len(params.shape) == 1:
        # biases
        nn.init.zeros_(params)
    else:
        raise ValueError("Unknown parameter shape: {}".format(params.shape))
  
with loss.value_network_params.to_module(loss.value_network):
    loss.apply(lambda x: reset_parameters(x.data) if hasattr(x, "data") else None)
```

And for handling the target_network_params I think we could simply do something like:
```python
loss.target_value_network_params.update(loss.value_network_params)
```

What do you think? I think we can close the draft. But we might want to mention the way to reset parameters somewhere in the docs.

**Comment by vmoens** [comment]
> loss.target_value_network_params.update(loss.value_network_params)

This won't work because the target params are locked (you can't update them). They're locked because we want to avoid this kind of operation :) 
You should update the data inplace:
```python
loss.target_value_network_params.apply(lambda dest, src: dest.data.copy_(src), loss.value_network_params)
```

**Comment by vmoens** [comment]
```python
def reset_parameters(params):
    """ User specified resetting function depending on their needs for initialization """
    if len(params.shape) > 1:
        # weights
        nn.init.xavier_uniform_(params)
    elif len(params.shape) == 1:
        # biases
        nn.init.zeros_(params)
    else:
        raise ValueError("Unknown parameter shape: {}".format(params.shape))
  
with loss.value_network_params.to_module(loss.value_network):
    loss.apply(lambda x: reset_parameters(x.data) if hasattr(x, "data") else None)
```
Unfortunately this isn't very generic
(1) all tensors have a data attribute, even buffers. By doing this you will also use Xavier init on batch-norm buffers if they're 2d
(2) If the model has a mixture of linear, conv and other layers it's going to be hard to have a fine grained control over the params being updated.

Not all modules are "weights" and "biases" and "biases" can be 2d (my point is: the dimension is a very indirect determinator of the tensor role in a model)

The way I usually see this work is to use the module `reset_parameters` if there is one, which provides a better control over difference in initialization methods. 

Maybe we could allow the user to pass a reset function, but in that case we don't even need to re-populate the module (we can just do `tensordict.apply(reset)`). Note that you could also do
```python
def reset(name, tensor):
    if name == "bias":
        tensor.data.zero_()
    if name == "weight":
        nn.init.xavier_uniform_(tensor)
tensordict.apply(reset, named=True)
```
which is more straightforward IMO

**Comment by gtnv** [comment]
hey @BY571, is this still something you're working on? if not, should we close it?
